scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala

```diff
@@ -296,6 +296,65 @@ object TreeSyntax {
       }
       !isOnlyChildOfOnlyChild(t)
     }
+    def guessNeedsLineSep(t: Template): Boolean =
+      t.stats.isEmpty
+    @annotation.tailrec
+    def guessNeedsLineSep(t: Stat): Boolean = t match {
+      case _: Decl.Val => true
+      case _: Decl.Var => true
+      case _: Decl.Def => true
+      case _: Decl.Type => false
+      case Defn.Val(_, _, _, rhs) => guessNeedsLineSep(rhs)
+      case Defn.Var.After_4_7_2(_, _, _, body) => guessNeedsLineSep(body)
+      case Defn.Def.After_4_7_3(_, _, _, _, body) => guessNeedsLineSep(body)
+      case _: Defn.Macro => true
+      case _: Defn.Type => true
+      case Defn.Class.After_4_6_0(_, _, _, _, template) => guessNeedsLineSep(template)
+      case Defn.Trait.After_4_6_0(_, _, _, _, template) => guessNeedsLineSep(template)
+      case Defn.Object(_, _, template) => guessNeedsLineSep(template)
+      case _: Term.This => true
+      case _: Term.Super => unreachable
+      case _: Term.Name => true
+      case _: Term.Select => true
+      case Term.ApplyUnary(_, arg) => guessNeedsLineSep(arg)
+      case _: Term.Apply => true
+      case _: Term.ApplyType => true
+      case Term.ApplyInfix.After_4_6_0(_, _, _, Term.ArgClause(term :: Nil, _)) =>
+        guessNeedsLineSep(term)
+      case _: Term.ApplyInfix => true
+      case Term.Assign(_, rhs) => guessNeedsLineSep(rhs)
+      case Term.Return(expr) => guessNeedsLineSep(expr)
+      case Term.Throw(expr) => guessNeedsLineSep(expr)
+      case _: Term.Ascribe => true
+      case _: Term.Annotate => false
+      case _: Term.Tuple => true
+      case _: Term.Block => false
+      case Term.If.After_4_4_0(_, thenp, Lit.Unit(), _) => guessNeedsLineSep(thenp)
+      case Term.If.After_4_4_0(_, _, elsep, _) => guessNeedsLineSep(elsep)
+      case _: Term.Match => false
+      case Term.Try(_, _, Some(term)) => guessNeedsLineSep(term)
+      case Term.Try(_, _ :: _, _) => false
+      case Term.Try(expr, Nil, None) => guessNeedsLineSep(expr)
+      case Term.TryWithHandler(_, _, Some(term)) => guessNeedsLineSep(term)
+      case Term.TryWithHandler(_, term, _) => guessNeedsLineSep(term)
+      case Term.Function.After_4_6_0(_, body) => guessNeedsLineSep(body)
+      case _: Term.PartialFunction => false
+      case Term.While(_, body) => guessNeedsLineSep(body)
+      case _: Term.Do => false
+      case Term.For(_, body) => guessNeedsLineSep(body)
+      case Term.ForYield(_, body) => guessNeedsLineSep(body)
+      case _: Term.New => true
+      case _: Term.NewAnonymous => false
+      case _: Term.Placeholder => unreachable
+      case _: Term.Eta => false
+      case _: Term.Repeated => false
+      case _: Term.Param => unreachable
+      case _: Term.Interpolate => true
+      case _: Term.Xml => true
+      case _: Import => false
+      case _: Lit => true
```

Comment:
why here?

Reply:
Literal + \n + { expr } is an apply, so
```scala
"foo"
{ 0 }
```
is `"foo"(0) == 'f'`, so it requires a new line here.